translator/trace/jaeger/jaegerproto_to_traces.go

```diff
@@ -0,0 +1,271 @@
+	jSpans := batch.GetSpans()
+
+	if jProcess == nil && len(jSpans) == 0 {
+		return traceData
+	}
+
+	rss := traceData.ResourceSpans()
+	rss.Resize(1)
+	rs := rss.At(0)
+	jProcessToInternalResource(jProcess, rs.Resource())
+
+	if len(jSpans) == 0 {
+		return traceData
+	}
+
+	ilss := rs.InstrumentationLibrarySpans()
+	ilss.Resize(1)
+	jSpansToInternal(jSpans, ilss.At(0).Spans())
+
+	return traceData
+}
+
+func jProcessToInternalResource(process *model.Process, dest pdata.Resource) {
+	if process == nil {
+		return
+	}
+
+	dest.InitEmpty()
+
+	serviceName := process.GetServiceName()
+	tags := process.GetTags()
+	if serviceName == "" && tags == nil {
+		return
+	}
+
+	attrs := dest.Attributes()
+
+	if serviceName != "" {
+		attrs.UpsertString(conventions.AttributeServiceName, serviceName)
+	}
+
+	for _, tag := range tags {
+
+		// Handle special keys in the tags.
+		switch tag.GetKey() {
+		case "hostname":
+			attrs.UpsertString(conventions.AttributeHostHostname, tag.GetVStr())
+			continue
+		case "jaeger.version":
+			attrs.UpsertString(conventions.OCAttributeExporterVersion, "Jaeger-"+tag.GetVStr())
+			continue
+		}
+
+		switch tag.GetVType() {
+		case model.ValueType_STRING:
+			attrs.UpsertString(tag.Key, tag.GetVStr())
+		case model.ValueType_BOOL:
+			attrs.UpsertBool(tag.Key, tag.GetVBool())
+		case model.ValueType_INT64:
+			attrs.UpsertInt(tag.Key, tag.GetVInt64())
+		case model.ValueType_FLOAT64:
+			attrs.UpsertDouble(tag.Key, tag.GetVFloat64())
+		case model.ValueType_BINARY:
+			attrs.UpsertString(tag.Key, base64.StdEncoding.EncodeToString(tag.GetVBinary()))
+		default:
+			attrs.UpsertString(tag.Key, fmt.Sprintf("<Unknown Jaeger TagType %q>", tag.GetVType()))
+		}
+	}
+}
+
+func jSpansToInternal(spans []*model.Span, dest pdata.SpanSlice) {
+	if len(spans) == 0 {
+		return
+	}
+
+	dest.Resize(len(spans))
+	i := 0
+	for _, span := range spans {
+		if span == nil || reflect.DeepEqual(span, blankJaegerProtoSpan) {
+			continue
+		}
+		jSpanToInternal(span, dest.At(i))
+		i++
+	}
+
+	if i < len(spans) {
+		dest.Resize(i)
+	}
+}
+
+func jSpanToInternal(span *model.Span, dest pdata.Span) {
+	dest.SetTraceID(pdata.TraceID(tracetranslator.UInt64ToByteTraceID(span.TraceID.High, span.TraceID.Low)))
+	dest.SetSpanID(pdata.SpanID(tracetranslator.UInt64ToByteSpanID(uint64(span.SpanID))))
+	dest.SetName(span.OperationName)
+	dest.SetStartTime(pdata.TimestampUnixNano(uint64(span.StartTime.UnixNano())))
+	dest.SetEndTime(pdata.TimestampUnixNano(uint64(span.StartTime.Add(span.Duration).UnixNano())))
+
+	parentSpanID := uint64(span.ParentSpanID())
+	if parentSpanID != 0 {
+		dest.SetParentSpanID(pdata.SpanID(tracetranslator.UInt64ToByteSpanID(parentSpanID)))
+	}
+
+	attrs := jTagsToInternalAttributes(span.Tags)
+	setInternalSpanStatus(attrs, dest.Status())
+	if spanKindAttr, ok := attrs[tracetranslator.TagSpanKind]; ok {
+		dest.SetKind(jSpanKindToInternal(spanKindAttr.StringVal()))
+	}
+	dest.Attributes().InitFromMap(attrs)
```

Comment:
We convert a slice to map here, then convert the map to slice again. I suspect this may be slow. Let's get a perf profile once this is merged and see what it shows. If it is slow we can probably just do a one pass and convert the source Tags slice directly to dest attributes.
This is OK for now, but let's verify it.